### PR TITLE
Fix dimensions of kernel evaluation of changepoint kernel

### LINKED ISSUE #1440: ChangePoints kernel breaks for len(X) != len(X2)
# Bug 
There seems to be a bug for models using the Changepoints kernel, whereby given a model 
m =gpflow.models.GPR(data=(X,y), kernel=k, mean_function=None)
with k a Changepoint kernel, (using a combination of any base kernels), the model won't produce a value for predict_y. 

## To reproduce

**Minimal, reproducible example**

```python
import numpy as np
import gpflow
X = np.linspace(0,100,100).reshape(100,1)
base_k1 = gpflow.kernels.Matern32(lengthscales=0.2)
base_k2 = gpflow.kernels.Matern32(lengthscales=2.0)
k = gpflow.kernels.ChangePoints([base_k1, base_k2], [0.0], steepness=5.0)
k(X)  # works

N = 25  # anything other than N=100 will reproduce the bug
xx = np.linspace(0,50,N).reshape(N,1)
k(X, xx)  # breaks
```

**Stack trace, or error message**
```
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-25-d1dbc7941bae> in <module>
----> 1 k(X, xx)

~/Code/GPflow/gpflow/kernels/base.py in __call__(self, X, X2, full_cov, presliced)
    170 
    171         else:
--> 172             return self.K(X, X2)
    173 
    174     def __add__(self, other):

~/Code/GPflow/gpflow/kernels/changepoints.py in K(self, X, X2)
     83         N = tf.shape(X)[0]
     84         ones = tf.ones((N, N, 1), dtype=X.dtype)
---> 85         starters = tf.concat([ones, starters], axis=2)
     86         stoppers = tf.concat([stoppers, ones], axis=2)
     87 

~/anaconda3/envs/gpflux2/lib/python3.7/site-packages/tensorflow_core/python/util/dispatch.py in wrapper(*args, **kwargs)
    178     """Call target, and fall back on dispatchers if there is a TypeError."""
    179     try:
--> 180       return target(*args, **kwargs)
    181     except (TypeError, ValueError):
    182       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/anaconda3/envs/gpflux2/lib/python3.7/site-packages/tensorflow_core/python/ops/array_ops.py in concat(values, axis, name)
   1515           dtype=dtypes.int32).get_shape().assert_has_rank(0)
   1516       return identity(values[0], name=name)
-> 1517   return gen_array_ops.concat_v2(values=values, axis=axis, name=name)
   1518 
   1519 

~/anaconda3/envs/gpflux2/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_array_ops.py in concat_v2(values, axis, name)
   1116         pass  # Add nodes to the TensorFlow graph.
   1117     except _core._NotOkStatusException as e:
-> 1118       _ops.raise_from_not_ok_status(e, name)
   1119   # Add nodes to the TensorFlow graph.
   1120   if not isinstance(values, (list, tuple)):

~/anaconda3/envs/gpflux2/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py in raise_from_not_ok_status(e, name)
   6604   message = e.message + (" name: " + name if name is not None else "")
   6605   # pylint: disable=protected-access
-> 6606   six.raise_from(core._status_to_exception(e.code, message), None)
   6607   # pylint: enable=protected-access
   6608 

~/anaconda3/envs/gpflux2/lib/python3.7/site-packages/six.py in raise_from(value, from_value)

InvalidArgumentError: ConcatOp : Dimensions of inputs should match: shape[0] = [100,100,1] vs. shape[1] = [100,25,1] [Op:ConcatV2] name: concat
```

## Expected behavior

Should return the kernel matrix of shape (100, 25).

## System information

* GPflow version: 2.0.1 (reproduced on `develop`)
* GPflow installed from: 'pip install gpflow'
* TensorFlow version: 2.1.0
* Python version  3.7.5
* Operating system Ubuntu Linux (18.04.3)

## Additional context

For me changing these two lines in changepoints.py or kernels did the trick: 

```python
    def K(self, X: tf.Tensor, X2: Optional[tf.Tensor] = None) -> tf.Tensor:
        sig_X = self._sigmoids(X)  # N x 1 x Ncp
        sig_X2 = self._sigmoids(X2) if X2 is not None else sig_X

        # `starters` are the sigmoids going from 0 -> 1, whilst `stoppers` go
        # from 1 -> 0, dimensions are N x N x Ncp
        starters = sig_X * tf.transpose(sig_X2, perm=(1, 0, 2))
        stoppers = (1 - sig_X) * tf.transpose((1 - sig_X2), perm=(1, 0, 2))

        # prepend `starters` with ones and append ones to `stoppers` since the
        # first kernel has no start and the last kernel has no end
        N = tf.shape(X)[0]
        M = tf.shape(X2)[0] if X2 is not None else N # THIS IS THE FIX
        ones = tf.ones((N, M, 1), dtype=X.dtype) #PREVIOUSLY N WAS IN PLACE OF M HERE
        starters = tf.concat([ones, starters], axis=2)
        stoppers = tf.concat([stoppers, ones], axis=2)
```

However I have not run any kind of tests to make sure this would not break anything else (neither have I added a unit test for the changepoint kernel). It could be that I am using the changepoint kernel incorrectly, in which case it would be very helpful if a short demo of using the kernel in a model could be appended to the changepoint kernel notebook :)

### PR DESCRIPTION
**Pull request by condnsdmatters**
Fixes #1440

### CODE REVIEW

File: tests/gpflow/kernels/test_changepoints.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    base_k1 = gpflow.kernels.Matern32(lengthscales=0.2)
+    base_k2 = gpflow.kernels.Matern32(lengthscales=2.0)
+    k = gpflow.kernels.ChangePoints([base_k1, base_k2], [0.0], steepness=5.0)
+    K = k(X)
+    assert K.shape == [100, 100]
+
+    N = 25
+    X2 = np.linspace(0, 50, N).reshape(N, 1)
+    K = k(X, X2)
+    assert K.shape == [100, 25]
```
Comment (st--) [inline tests/gpflow/kernels/test_changepoints.py:18]: re-use the N you've just defined? maybe call it N2 and do the same for N1=100?

File: tests/gpflow/kernels/test_changepoints.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
+import gpflow
+import numpy as np
+import gpflow
```
Comment (st--) [inline tests/gpflow/kernels/test_changepoints.py:4]: duplicated import
```suggestion
```

File: tests/gpflow/kernels/test_changepoints.py (hunk 1 above)
Comment (st--) [inline tests/gpflow/kernels/test_changepoints.py:18]:
```suggestion
    N = 100
    X = np.linspace(0, 100, N).reshape(N, 1)
    base_k1 = gpflow.kernels.Matern32(lengthscales=0.2)
    base_k2 = gpflow.kernels.Matern32(lengthscales=2.0)
    k = gpflow.kernels.ChangePoints([base_k1, base_k2], [0.0], steepness=5.0)
    K = k(X)
    assert K.shape == [N, N]

    N2 = 25
    X2 = np.linspace(0, 50, N2).reshape(N2, 1)
    K = k(X, X2)
    assert K.shape == [N, N2]
```

File: tests/gpflow/kernels/test_changepoints.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    assert K.shape == [100, 100]
+
+    N = 25
+    X2 = np.linspace(0, 50, N).reshape(N, 1)
+    K = k(X, X2)
+    assert K.shape == [100, 25]
+
+
+if __name__ == "__main__":
+    test_changepoint_with_X1_X2()
```
Comment (st--) [inline tests/gpflow/kernels/test_changepoints.py:22]:
```suggestion
```
I would not have this in the test file. Just run it with `pytest tests/gpflow/kernels/test_changepoints.py` - have one way to do it!

File: tests/gpflow/kernels/test_changepoints.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
+import gpflow
+import numpy as np
+import gpflow
+
+
+def test_changepoint_with_X1_X2():
+    X = np.linspace(0, 100, 100).reshape(100, 1)
```
Comment (condnsdmatters) [inline tests/gpflow/kernels/test_changepoints.py:8]:
```suggestion
    N1 = 100
    X = np.linspace(0, 100, N1).reshape(N1, 1)
```

File: tests/gpflow/kernels/test_changepoints.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    base_k1 = gpflow.kernels.Matern32(lengthscales=0.2)
+    base_k2 = gpflow.kernels.Matern32(lengthscales=2.0)
+    k = gpflow.kernels.ChangePoints([base_k1, base_k2], [0.0], steepness=5.0)
+    K = k(X)
+    assert K.shape == [100, 100]
+
+    N = 25
```
Comment (condnsdmatters) [inline tests/gpflow/kernels/test_changepoints.py:15]:
```suggestion
    N2 = 25
```

File: tests/gpflow/kernels/test_changepoints.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    X2 = np.linspace(0, 50, N).reshape(N, 1)
```
Comment (condnsdmatters) [inline tests/gpflow/kernels/test_changepoints.py:16]:
```suggestion
    X2 = np.linspace(0, 50, N2).reshape(N2, 1)
```

**Pull request merged**